man/man1/chacl.1
.TH chacl 1
.SH NAME
chacl \- change or list the access control list of a file or directory
.SH SYNOPSIS
.B chacl acl pathname...
.br
.B chacl -b acl dacl pathname...
.br
.B chacl -d dacl pathname...
.br
.B chacl -R pathname...
.br
.B chacl -D pathname...
.br
.B chacl -B pathname...
.br
.B chacl -l pathname...
.br
.SH DESCRIPTION
.I chacl\^
changes or lists the Access Control List(s) for a file or directory.
The \f4-b\f1 flag indicates that there are two ACLs to change, 
the first is the
file access ACL and the second the directory default ACL. The \f4-d\f1 flag is 
used to set only the default ACL of a directory.  
The ACL(s) specified are applied to each file in the \f4pathname\f1 arguments.
The \f4-R\f1 and \f4-D\f1
flags are used to remove the file access ACL and directory default ACL respectively,
the \f4-B\f1 flag will remove all ACLs. 
The \f4-l\f1 flag is used to list the access ACL and possibly the
default ACL associated with the specified files or directories.
.P
Each ACL is a string which is processed via \f4acl_from_text\fP(3).
These strings are made up of comma separated clauses each of which
is of the form, tag:name:perm.  Where \f4tag\fP can be
.TP
"user" (or "u")
indicating that the entry is a user ACL entry.
.TP
"group" (or "g")
indicating that the entry is a group ACL entry.
.TP
"other" (or "o")
indicating that the entry is an other ACL entry.
.TP
"mask" (or "m")
indicating that the entry is a mask ACL entry.
.P
\f4name\fP is a string which is the user or group name for the ACL entry.
A null \f4name\fP in a user or group ACL entry indicates the file's
owner or file's group.
\f4perm\fP is the string "rwx" where each of the entries may be replaced
by a "-" indicating no access of that type, e.g. "r-x", "--x", "---".
.SH EXAMPLES

A minimum ACL:
\f3chacl u::rwx,g::r-x,o::r-- file\fP
.br
The file ACL is set so that the file's owner has rwx, the file's
group has read and execute, and others have read only access to the file.
.P
An ACL that is not a minimum ACL, that is, one that specifies
a user or group other than the file's owner or owner's group,
must contain a mask entry:
\f4chacl u::rwx,g::r-x,o::r--,u:guest:r--,m::r-x file1 file2\fP
.P
To set the default and access ACLs on newdir to be the 
same as on oldir, you could type:
.P
\f4chacl -b `chacl -l olddir | sed -e 's/.*\\[//' -e 's#/# #' -e 's/]$//'` newdir
\fP
.SH CAUTIONS
\f4chacl\fP can replace the existing ACL.  To add or delete entries, you
must first do \f4chacl -l\fP to get the existing ACL, and use the output
to form the arguments to \f4chacl\fP.
.P

Changing the permission bits of a file will change the file access ACL settings 
(see \f4chmod\fP(1)). However, file-creation mode masks (\f4umask\fP(1)) will
not affect the access ACL settings of files created using directory 
default ACLs. 
.P
ACLs are filesystem extended attributes and are not typically archived 
or restored using conventional archiving utilities.
See \f4attr\fP(1) for more information 
about Extended Attributes and see \f4xfsdump\fP(8) 
for a method to back them up.
.SH SEE ALSO
acl_from_text(3),
umask(1),
xfsdump(8).
.SH CAVEATS
This command is likely to be deprecated when a standard file-system
independent ACL library is available.


doc/CHANGES
2.0.6 (08 April 2002)
* Fix a bug that causes setfacl not to report error messages if
  acl-set_file() fails (the bug was introduced in version 2.0.3).

2.0.5 (26 March 2002)
* Fix up acl_to_text compliance with POSIX draft.
* Add the new acl(5) manual page.
* Add a manual section 3 directory and Makefile.
* Add the libacl manual pages.
* Rename the existing acl_to_any_text function to __acl_to_any_text, and
  add a new acl_to_any_text function that only contains parameters that
  are actually useful.  Make acl_to_text and acl_to_any_text invoke
  __acl_to_any_text internally.  Update all uses of acl_to_any_text in
  getfacl/getfacl.c, setfacl/do_set.c, and chacl/chacl.c.
* Move the acl_size function form the LIBACL_CFILES to the POSIX_CFILES
  section (it is a POSIX function)
* Rearrange libacl routines so that several non-POSIX routines which
  were being exported are no longer exported (eg. acl_entry_to_any_str,
  acl_get_file_mode, acl_get_fd_mode, acl_set_file_mode, acl_set_fd_mode).

2.0.4 (12 March 2002)
* Man page updates from Andreas
* Test script updates from Andreas

2.0.3 (05 March 2002)
* Clean up the --default option handling in setfacl. The old
  workarounds caused a bug for unusual input.
* Changes to the --test output format setfacl generates: ACLs that
  are not changed are now displayed as `*'.
* Fix a bug in setfacl/sequence.c:seq_delete_cmd().
* Minor changes to test scripts

2.0.2 (01 March 2002)
* Apply several patches from Andreas, namely:
* man page fixes
* libacl code reformatting
* acl_from_text errno handling

2.0.1 (27 February 2002)
* Apply Andreas Gruenbacher's diffs.
* Fix up chacl for deletion of access ACL to be in line with Andreas. 
* Incorporate the Debian packaging again.

2.0.0 (21 February 2002)
* Reworked to use the new official system call API.
* Sync up with the XFS project, the SGI folk now use this source.
* Jumped to version 2 to allow XFS users to upgrade
  (Rationale: the XFS ACL user tools were at version 1.1.X, and
  packaging tools like rpm, dpkg, etc. must be presented with a
  greater version number to allow an upgrade to proceed).
* Added the chacl command to ease migration for existing XFS users,
  and for compatibility with IRIX.
* Added a flag to allow acl_print to produce a single-line ACL, in
  addition to the multi-line format.
* Extended attribute documentation has moved into the extended
  attribute package from SGI ("attr"), this ACL package now deals
  exclusively with ACLs.
* acl_from_text sometimes did not set errno when failing.
* Moved files and simplified #includes in libacl

0.7.22
* Create a shared version of libacl, instead of linking statically.
* Added missing error reporting to getfacl do_print() function.     

0.7.16
* Minor cleanups/fixes in the make files
* The test cases in test/src have been updated to match the
  corrected ACL entry syntax (see the 0.7.15 changes).

0.7.15
* acl_from_text() now accepts mask and other entries with either two
  or three fields (`other:rwx' or `other::rwx'). Posix 1003.1e specifies
  that these entries should have three entries, while under Solaris,
  they have only two fields.
* setfacl now parses mask and other entries with two or three entries.
* getfacl now produces mask and other entries with three fields
  (`mask::rwx' instead of the previous `mask:rwx').
* acl_get_fd(), acl_get_file(), acl_set_fd(), acl_set_file() now only
  fail with errno=ENOSYS if the underlying filesystem doesn't support
  ACLs. acl_get_fd() and acl_get_file() construct ACLs from the file
  mode permission bits if the kernel supports ACLs, but no ACL extended
  attributes are defined for the file (errno=ENOATTR).
* acl_to_text() wrongly separated ACL entires with ',' instead of '\n'.
* acl_from_text() only accepted three-letter permission strings
  (rwx, ---, rw-, etc.). Now also short strings (rw, -, etc.) are
  supported.

0.7.10
* acl_get_fd(), acl_get_file() now don't return basic ACL's anymore.
* acl_set_fd(), acl_set_file() fail on filesystems that don't support
  ACLs, even if the ACL to be set is a basice ACL.
* If msgfmt is not present on a system, the message catalog(s) are
  no longer being built.

0.7.8 (19 February 2001)
* Buffer overflow bug in ACL library (acl_to_any_text).
  (Reported by Charles Bertsch <CBertsch@microtest.com>)

0.7.2 (22 November 2000)
* Replace rman with groff for building HTML versions of the
  manual pages.
* Add descriptions of the permissions required for manipulating
  ACLs to the manual pages.
* Add long options to getfacl and setfacl, update the manual pages.
* Remove some small bugs from the src/showacl Perl script.
* showacl is now integrated in getfacl.
* Bug in acl_to_any_text() for ACLs with 0 entries fixed.
* Bug in acl_to_text(): Used '\n' as the ACL entry separator.
* German messages updated.

0.7.1 (23 October 2000)
* Basic GNU gettext support and a German translation of messages
  added.

0.7.0 (22 October 2000)
* Updates to the documentation
* Now an ACL_MASK entry is cloned from the ACL_GROUP_OBJ entry
  when needed also when the ACL_MASK entry is not re-calculated
  (-n option), so setfacl doesn't complain about an invalid ACL.

0.6.8 (11 October 2000)
* src/walk_tree.c was stat()ing the targets of symbolic links in
  WALK_PHYSICAL mode.
* Removed spurious definitions from src/gettext.h that caused
  problems with several versions of GNU gettext headers
* setfacl calls acl_delete_def_file() if removing the default ACL
  of a directory that doesn't have a default ACL. i(If the user lacks
  appropriate permissions, that no-operation fails.) Fixed.
* Tiny bug in src/showacl script (mask entry not applied to owning
  group entry)

0.6.5 (10 September 2000)
* Command line options revised for POSIX compatibility (now defaults to
  -H half-logical walk, added -L logical walk, -P physical walk).
* Funny post-order option no longer documented.

0.6.0 (24 March 2000)
* Permissions were not set to the union of the ACL_GROUP_OBJ and

chacl/chacl.c
/*
 * Copyright (c) 2001-2002 Silicon Graphics, Inc.  All Rights Reserved.
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 * 
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc., 59
 * Temple Place - Suite 330, Boston MA 02111-1307, USA.
 * 
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 * 
 * http://www.sgi.com 
 * 
 * For further information regarding this notice, see: 
 * 
 * http://oss.sgi.com/projects/GenInfo/SGIGPLNoticeExplan/
 */

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <libgen.h>
#include <stdio.h>
#include <errno.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/acl.h>
#include <acl/libacl.h>

#include <locale.h>
#include <libintl.h>
#define _(String) gettext (String)

static int acl_delete_file (const char * path, acl_type_t type);
static int list_acl(char *file);
static int set_acl(acl_t acl, acl_t dacl, const char *fname);
static int walk_dir(acl_t acl, acl_t dacl, const char *fname);

static char *program;
static int rflag;

static void
usage(void)
{
	fprintf(stderr, _("Usage:\n"));
	fprintf(stderr, _("\t%s acl pathname...\n"), program);
	fprintf(stderr, _("\t%s -b acl dacl pathname...\n"), program);
	fprintf(stderr, _("\t%s -d dacl pathname...\n"), program);
	fprintf(stderr, _("\t%s -R pathname...\n"), program);
	fprintf(stderr, _("\t%s -D pathname...\n"), program);
	fprintf(stderr, _("\t%s -B pathname...\n"), program);
	fprintf(stderr, _("\t%s -l pathname...\t[not IRIX compatible]\n"),
			program);
	fprintf(stderr, _("\t%s -r pathname...\t[not IRIX compatible]\n"),
			program);
	exit(1);
}

int
main(int argc, char *argv[])
{
	char *file;
	int switch_flag = 0;            /* ensure only one switch is used */
	int args_required = 1;	
	int failed = 0;			/* exit status */
	int c;				/* For use by getopt(3) */
	int dflag = 0;			/* a Default ACL is desired */
	int bflag = 0;			/* a both ACLs are desired */
	int Rflag = 0;			/* set to true to remove an acl */
	int Dflag = 0;			/* set to true to remove default acls */
	int Bflag = 0;			/* set to true to remove both acls */
	int lflag = 0;			/* set to true to list acls */
	acl_t acl = NULL;		/* File ACL */
	acl_t dacl = NULL;		/* Directory Default ACL */

	program = basename(argv[0]);

	/* parse arguments */
	while ((c = getopt(argc, argv, "bdlRDBr")) != -1) {
		if (switch_flag) 
			usage();
		switch_flag = 1;

		switch (c) {
			case 'b':
				bflag = 1;
				args_required = 3;
				break;
			case 'd':
				dflag = 1;
				args_required = 2;
				break;
			case 'R':
				Rflag = 1;
				break;
			case 'D':
				Dflag = 1;
				break;
			case 'B':
				Bflag = 1;
				break;
			case 'l':
				lflag = 1;
				break;
			case 'r':
				rflag = 1;
				break;
			default:
				usage();
				break;
		}
	}

	/* if not enough arguments quit */
	if ((argc - optind) < args_required)
		usage();

        /* list the acls */
	if (lflag) {
		for (; optind < argc; optind++) {	
			file = argv[optind];
			if (!list_acl(file))
				failed++;
		}
		return(failed);
	}

	/* remove the acls */
	if (Rflag || Dflag || Bflag) {
		for (; optind < argc; optind++) {	
			file = argv[optind];
			if (!Dflag &&
			    (acl_delete_file(file, ACL_TYPE_ACCESS) == -1)) {
				fprintf(stderr,
			"%s: error removing access acl on \"%s\": %s\n",
					program, file, strerror(errno));
				failed++;
			}
			if (!Rflag &&
			    (acl_delete_file(file, ACL_TYPE_DEFAULT) == -1)) {
				fprintf(stderr,
			_("%s: error removing default acl on \"%s\": %s\n"),
					program, file, strerror(errno));
				failed++;
			}
		}
		return(failed);
	}

	/* file access acl */
	if (! dflag) { 
		acl = acl_from_text(argv[optind]);
		failed = acl_check(acl, &c);
		if (failed < 0) {
			fprintf(stderr, _("%s: %s - %s\n"),
				program, argv[optind], strerror(errno));
			return 1;
		}
		else if (failed > 0) {
			fprintf(stderr,
				_("%s: access ACL '%s': %s at entry %d\n"),
				program, argv[optind], acl_error(failed), c);
			return 1;
		}
		optind++;
	}


	/* directory default acl */
	if (bflag || dflag) {
		dacl = acl_from_text(argv[optind]);
		failed = acl_check(dacl, &c);
		if (failed < 0) {
			fprintf(stderr, _("%s: %s - %s\n"),
				program, argv[optind], strerror(errno));
			return 1;
		}
		else if (failed > 0) {
			fprintf(stderr,
				_("%s: access ACL '%s': %s at entry %d\n"),
				program, argv[optind], acl_error(failed), c);
			return 1;
		}
		optind++;
	}

	/* place acls on files */
	for (; optind < argc; optind++)
		failed += set_acl(acl, dacl, argv[optind]);

	if (acl)
		acl_free(acl);
	if (dacl)
		acl_free(dacl);

	return(failed);
}

/* 
 *   deletes an access acl or directory default acl if one exists
 */ 
static int 
acl_delete_file(const char *path, acl_type_t type)
{
	int error = 0;

	/* converts access ACL to a minimal ACL */
	if (type == ACL_TYPE_ACCESS) {
		acl_t acl;
		acl_entry_t entry;
		acl_tag_t tag;

		acl = acl_get_file(path, ACL_TYPE_ACCESS);
		if (!acl)
			return -1;
		error = acl_get_entry(acl, ACL_FIRST_ENTRY, &entry);
		while (error == 1) {
			acl_get_tag_type(entry, &tag);
			switch(tag) {
				case ACL_USER:
				case ACL_GROUP:
				case ACL_MASK:
					acl_delete_entry(acl, entry);
					break;
			 }
			error = acl_get_entry(acl, ACL_NEXT_ENTRY, &entry);
		}
		if (!error)
			error = acl_set_file(path, ACL_TYPE_ACCESS, acl);
	} else
		error = acl_delete_def_file(path);
	return(error);
}

/*
 *    lists the acl for a file/dir in short text form
 *    return 0 on failure
 *    return 1 on success
 */
static int
list_acl(char *file)
{
	acl_t acl = NULL;
	acl_t dacl = NULL;
	char *acl_text, *dacl_text = NULL;

	if ((acl = acl_get_file(file, ACL_TYPE_ACCESS)) == NULL) {
		fprintf(stderr, _("%s: cannot get access ACL on '%s': %s\n"),
			program, file, strerror(errno));
		return 0;
	}
	if ((dacl = acl_get_file(file, ACL_TYPE_DEFAULT)) == NULL &&
	    (errno != EACCES)) {	/* EACCES given if not a directory */
		fprintf(stderr, _("%s: cannot get default ACL on '%s': %s\n"),
			program, file, strerror(errno));
		return 0;
	}
	acl_text = acl_to_any_text(acl, NULL, ',', TEXT_ABBREVIATE);
	if (acl_text == NULL) {
		fprintf(stderr, _("%s: cannot get access ACL text on '%s': %s\n"),
			program, file, strerror(errno));
		return 0;
	}
	if (acl_entries(dacl) > 0) {
		dacl_text = acl_to_any_text(dacl, NULL, ',', TEXT_ABBREVIATE);
		if (dacl_text == NULL) {
			fprintf(stderr, _("%s: cannot get default ACL text on '%s': %s\n"),
				program, file, strerror(errno));
			return 0;
		}
	}
	if (dacl_text) {
		printf("%s [%s/%s]\n", file, acl_text, dacl_text);
		acl_free(dacl_text);
	} else
		printf("%s [%s]\n", file, acl_text);
	acl_free(acl_text);
	acl_free(acl);
	acl_free(dacl);
	return 1;
}

static int
set_acl(acl_t acl, acl_t dacl, const char *fname)
{
	int failed = 0;

	if (rflag)
		failed += walk_dir(acl, dacl, fname);

	/* set regular acl */
	if (acl && acl_set_file(fname, ACL_TYPE_ACCESS, acl) == -1) {
		fprintf(stderr, _("%s: cannot set access acl on \"%s\": %s\n"),
			 program, fname, strerror(errno));
		failed++;
	}
	/* set default acl */
	if (dacl && acl_set_file(fname, ACL_TYPE_DEFAULT, dacl) == -1) {
		fprintf(stderr, _("%s: cannot set default acl on \"%s\": %s\n"),
			program, fname, strerror(errno));
		failed++;
	}

	return(failed);
}

static int
walk_dir(acl_t acl, acl_t dacl, const char *fname)
{
	int failed = 0;
	DIR *dir;
	struct dirent64 *d;
	char *name;

	if ((dir = opendir(fname)) == NULL) {
		if (errno != ENOTDIR) {
			fprintf(stderr, _("%s: opendir failed: %s\n"),
				program, strerror(errno));
			return(1);
		}
		return(0);	/* got a file, not an error */
	}

	while ((d = readdir64(dir)) != NULL) {
		/* skip "." and ".." entries */
		if (strcmp(d->d_name, ".") == 0 || strcmp(d->d_name, "..") == 0)
			continue;
		
		name = malloc(strlen(fname) + strlen(d->d_name) + 2);
		if (name == NULL) {
			fprintf(stderr, _("%s: malloc failed: %s\n"),
				program, strerror(errno));
			exit(1);
		}
		sprintf(name, "%s/%s", fname, d->d_name);

		failed += set_acl(acl, dacl, name);
		free(name);
	}
	closedir(dir);

	return(failed);
}
